### PR TITLE
Rename useConnection to useShowMorePagination

### PR DESCRIPTION
**Pull request by philipp-spiess**
Part of https://github.com/sourcegraph/sourcegraph/issues/44706

From [Slack](https://sourcegraph.slack.com/archives/C04931KQVRC/p1669050551421889):

> We’re working on a replacement for useConnection with better defaults and support for bi-directional pagination. The existing useConnection was added as a replacement for <FilteredConnection/> a bit over a year ago and is a drop-in replacement for the older pattern on the frontend (without having to change anything in the backend). It does this very well: It’s currently simple to move from <FilteredConnection/> to the new hook implementation.
>
> As part of the strategic scale-testing, I’m currently working with @Naman and @Kelli Rockwell to take this one step further: We’re rethinking how we’re doing pagination on the backend as well. This means that we now recommend a fully cursor-based implementation in the backend that supports traversal in both directions and we’re also changing the UI default from being an append-only list to showing individual pages. This will need new front-end components: the hook above and a new pagination UI component.
>
> We want all new paginations to be using this new pattern (a more detailed writeup is following shortly as well) so it should be clear that the existing useConnection is no longer the recommended default. However it’s not easy to migrate from useConnection to the new hook (because it requires backend support and for some views it’s maybe not what we want either) so I assume that both patterns will coexist.

This is the first step to my plan: Rename existing `useConnection` usages. @felixfbecker proposed the name `useShowMorePagination` which I find quite fitting. @quinnkeast suggested `useInfinitePagination` which would also work. Does anyone feel strongly about the name? 

## Test plan

This is just a large string replace, I’m relying on our tests to break if things are broken.

## App preview:

- [Web](https://sg-web-ps-rn-use-connection.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0.00 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/ps/rn-use-connection/compare-report.html;tab=live_object) for a full comparison between the commits 826ebae and 7b4c2a1 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by taras-yemets** [comment]
Another option could be to rename `useConnection` to `useFetchMore` (`fetchMore` is one of the functions returned by the hook) and use `usePagination` name for a hook returning prev/next page fetchers.

**Comment by philipp-spiess** [comment]
I like that but I’m not sure if pagination alone clear enough (we've discussed this with @felixfbecker in the [Slack thread](https://sourcegraph.slack.com/archives/C04931KQVRC/p1669050551421889)).

**Review by taras-yemets — APPROVED** [review]

File: client/web/src/components/FilteredConnection/hooks/useShowMorePagination.ts [hunk 1]
```diff
@@ -8,7 +8,7 @@ import { useSearchParameters, useInterval } from '@sourcegraph/wildcard'
 import { Connection, ConnectionQueryArguments } from '../ConnectionType'
 import { asGraphQLResult, hasNextPage, parseQueryInt } from '../utils'
 
-import { useConnectionUrl } from './useConnectionUrl'
+import { useShowMorePaginationUrl } from './useShowMorePaginationUrl'
 
 export interface UseConnectionResult<TData> {
```
Comment (taras-yemets) [inline client/web/src/components/FilteredConnection/hooks/useShowMorePagination.ts:13]: @philipp-spiess, what do you think about renaming this interface too?

**Comment by taras-yemets** [comment]
I'm still not sure about the naming.
We are going to have two `useSomeTypeOfPagination` hooks, and we are going to use these and derived hooks to get some kind of *connection* (e.g., [`useImportingChangesets`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@d1ade502bcc39061d56115666710138c03f86289/-/blob/client/web/src/enterprise/batches/batch-spec/edit/workspaces-preview/WorkspacesPreview.tsx?L96:43), [`useBulkOperationsListConnection`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@d1ade50/-/blob/client/web/src/enterprise/batches/detail/BulkOperationsTab.tsx?L37:68), [RepoRevisionSidebarSymbols](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@d1ade50/-/blob/client/web/src/repo/RepoRevisionSidebarSymbols.tsx?L135), etc.).
If I use some hook to get a _connection_ and some functions to manage it, I'd probably expect this hook to have word _'connection'_ in its name. If I use `useSomeTypeOfPagination` I'd expect to get a current page number, total pages, and get prev/next page functions (for example these `usePagination` examples: [1](https://www.algolia.com/doc/api-reference/widgets/pagination/react-hooks/#hook-usage), [2](https://github.com/erictooth/react-use-pagination#example)). 
Having `const { connection, error, loading, hasNextPage, fetchMore } = useShowMorePagination(...args)` with `connection` being the _'main'_ return value does not seem intuitive to me.

File: client/web/src/components/FilteredConnection/hooks/useShowMorePagination.ts (hunk 1 above)
Comment (philipp-spiess) [reply client/web/src/components/FilteredConnection/hooks/useShowMorePagination.ts]: @taras-yemets I forgot to push 🤦

**Comment by philipp-spiess** [comment]
@taras-yemets Why is the word `connection` so important in this case? E.g. cant `useBulkOperationsListConnection` also be called `useBulkOperationsListPagination`?

> If I use useSomeTypeOfPagination I'd expect to get a current page number, total pages, and get prev/next page functions 

In both cases you'd get a 'next page' function (`fetchMore` and `goToNextPage`/`goToPreviousPage`) but yeah we can't show you the current page number because that's not possible with cursor based navigation. I still think that there's not a fundamental difference between this hook, the [one I’m adding](https://github.com/sourcegraph/sourcegraph/pull/44664) and the ones you linked.

> Having const { connection, error, loading, hasNextPage, fetchMore } = useShowMorePagination(...args) with connection being the 'main' return value does not seem intuitive to me.

I think `connection` still makes sense as a term here because this is what GraphQL is calling the type of this thing. We can probably also return the current `nodes` instead because `connection` must have `nodes`, would that be better?

**Comment by taras-yemets** [comment]
@philipp-spiess, I don't have name suggestions I'm confident in that seem more suitable than the ones shared by folks in this PR and in the referenced thread.
I'm fine with the selected name.

**Comment by vdavid** [comment]
@philipp-spiess I'm not intimately familiar with the pagination effort, just saw this PR and was interested. I skimmed through the Slack convo and the related tracking issue, and other PRs, that's pretty much all the context I have. So just leaving my two cents here:

As far as I know, the term "connection" comes from the "GraphQL Cursor Connections" (I'll call it GCC in this message) standard: https://relay.dev/graphql/connections.htm
I guess the name `useConnection` wants to mean that "this is a hook that connects to a GraphQL API that follows the GCC standard, and returns a `connection` as defined in the GCC spec."

I understand @thenamankumar is working on back-end changes that improve our pagination. But I'm unsure: does it also mean that we're moving off the GCC standard?

a.) If **yes**, then renaming the hook to something like `useShowMorePagination` makes a lot of sense to me because it abstracts away GCC, and basically says that "this is a hook that can handle pagination with _some_ backend (that is not necessarily GCC-compatible)".
b.) If **no**, then I think keeping the term `connection` in the hook name is pretty elegant because it compresses a lot of information (a reference to the GCC standard) in a very short term.

Again, just my two cents, I think this rename can't hurt in any case, as it's very easy to change it later.

**Comment by philipp-spiess** [comment]
Thanks @vdavid!

> I understand @thenamankumar is working on back-end changes that improve our pagination. But I'm unsure: does it also mean that we're moving off the GCC standard?

**No!** The work that we're doing is to extend upon the GraphQL connection interface and add a helper in Go that makes it easier to: 1) uses _proper cursors_ (the current patterns all show lead to `OFFSET .. LIMIT ..` based SQL queries which defeats a lot of the purpose here) and 2) allows for forward _and backward_ navigation.

We still expect the GraphQL APIs to implement the Connection type.

> If no, then I think keeping the term connection in the hook name is pretty elegant because it compresses a lot of information (a reference to the GCC standard) in a very short term.

That is fair but I wonder how much additional value the knowledge that a hook uses an underlying GraphQL connection really brings: When adding a new usage of the hook, the types already make it clear that this is the required backend structure and when reading the code I feel like this term has just become a synonym for pagination in our codebase.

**Comment by lrhacker** [comment]
Given that we expect this to exist (for at least some period of time) in combination with the new `usePageSwitcherPagination` hook, it makes sense to me for this one to be named `useShowMorePagination`. For now, they both are clearly distinguishable about what they should be used for until we are able to support both cases with a single implementation.

We can provide more context/detail about the relevant implementation details (e.g. the use of a "connection") as needed/desired in the JSDoc comments for the hooks (and, like you mentioned @philipp-spiess, the type system also helps us describe this).

**Pull request merged**